wandb/util.py

```diff
@@ -19,35 +19,82 @@
 import six
 from six.moves import queue
 import textwrap
+from sys import getsizeof
+from collections import namedtuple
 
 import wandb
 from wandb import io_wrap
 from wandb import wandb_dir
 
 logger = logging.getLogger(__name__)
 
-# TODO: handle no numpy?
 try:
-    import numpy as np
     from torch import Tensor
 except ImportError:
     class Tensor(object):
         pass
 
+try:
+    from tensorflow import Tensor
+except ImportError:
+    class Tensor(object):
+        pass
+
+try:
+    import numpy as np
+except ImportError:
+    np = namedtuple('np', ['ndarray', 'generic'])
+
+try:
+    import pandas
+except ImportError:
+    pandas = namedtuple('pandas', ['DataFrame'])
+
 MAX_SLEEP_SECONDS = 60 * 5
+# TODO: Revisit these limits
+VALUE_BYTES_LIMIT = 100000
+NP_BYTES_LIMIT = 1000
+
+
+def fullname(o):
+    return o.__class__.__module__ + "." + o.__class__.__name__
+
+
+def json_friendly(obj):
+    """Convert an object into something that's more becoming of JSON"""
+    converted = True
+    transformed = False
+    if isinstance(obj, (np.ndarray, Tensor, pandas.DataFrame)):
+        if getsizeof(obj) > NP_BYTES_LIMIT:
+            obj = {
+                "_type": fullname(obj),
+                "var": np.var(obj),
```

Comment:
note: these might not work since numpy is optionally imported

Reply:
I think we handle this because if I can’t import numpy then a create a fake numpy tuple that should never test positive and allow this code path to be reached.

Reply:
ah yeah it looks like pandas and tensorflow both have a direct dep on numpy, so this should always work